thunder/core/transforms.py

```diff
@@ -3702,7 +3702,10 @@ def _update_forward_with_new_saved_for_backward(forward_trace: Trace, saved_for_
         saved_for_backward (Sequence[Variable]): Saved_for_backward to use to
             update the forward trace.
     """
-    saved_for_backward = tree_map(lambda x: x.value if isinstance(x, NumberProxy) else x, saved_for_backward)
+    forward_trace_producers = utils.producers(forward_trace)
+    saved_for_backward = tree_map(
```

Comment:
nitpick: a comment explaining why we'd skip baking in compile time constant `x.value` would be great.

A TODO would be even better: Why we'd ever want to bake in number proxies and that's a future target that we should resolve.

Reply:
I agree, the comment on why is needed here. It's missing in a lot of places in Thunder in general...

This line of code is responsible of sending incorrect `saved_for_backward` to this function:
https://github.com/Lightning-AI/lightning-thunder/blob/067f15aae47ad71229732ca6c35a5d190135e48c/thunder/core/rematerialization.py#L643

There's another buggy usage https://github.com/Lightning-AI/lightning-thunder/blob/067f15aae47ad71229732ca6c35a5d190135e48c/thunder/core/rematerialization.py#L501

The fix can live inside `_update_forward_with_new_saved_for_backward` or `saved_for_backward` argument should be fixed at the callsites of `_update_forward_with_new_saved_for_backward`. Currently "saved_for_backward" is taken from the backward trace which has all inputs proxified and it's a mistake to put this information into the forward trace output because not all numbers are actually produced in forward.

Reply:
Let's move forward with this PR. I stand corrected, my [earlier comment](https://github.com/Lightning-AI/lightning-thunder/issues/541#issuecomment-2161521027) was mistaken :stuck_out_tongue: